sandbox/rocky/tf/algos/batch_polopt.py

```diff
@@ -103,35 +104,42 @@ def train(self, sess=None):
-                logger.log("Logging diagnostics...")
-                self.log_diagnostics(paths)
-                logger.log("Optimizing policy...")
-                self.optimize_policy(itr, samples_data)
-                logger.log("Saving snapshot...")
-                params = self.get_itr_snapshot(itr, samples_data)  # , **kwargs)
-                if self.store_paths:
-                    params["paths"] = samples_data["paths"]
-                logger.save_itr_params(itr, params)
-                logger.log("Saved")
-                logger.record_tabular('Time', time.time() - start_time)
-                logger.record_tabular('ItrTime', time.time() - itr_start_time)
-                logger.dump_tabular(with_prefix=False)
-                if self.plot:
-                    rollout(self.env, self.policy, animated=True, max_path_length=self.max_path_length)
-                    if self.pause_for_plot:
-                        input("Plotting evaluation run: Press Enter to "
-                              "continue...")
+        try:    
+            sess.run(tf.global_variables_initializer())
+            self.start_worker()
+            start_time = time.time()
+            initialize_plot = 0
+            for itr in range(self.start_itr, self.n_itr):
+                itr_start_time = time.time()
+                with logger.prefix('itr #%d | ' % itr):
+                    logger.log("Obtaining samples...")
+                    paths = self.obtain_samples(itr)
+                    logger.log("Processing samples...")
+                    samples_data = self.process_samples(itr, paths)
+                    logger.log("Logging diagnostics...")
+                    self.log_diagnostics(paths)
+                    logger.log("Optimizing policy...")
+                    self.optimize_policy(itr, samples_data)
+                    logger.log("Saving snapshot...")
+                    params = self.get_itr_snapshot(itr, samples_data)  # , **kwargs)
+                    if self.store_paths:
+                        params["paths"] = samples_data["paths"]
+                    logger.save_itr_params(itr, params)
+                    logger.log("Saved")
+                    logger.record_tabular('Time', time.time() - start_time)
+                    logger.record_tabular('ItrTime', time.time() - itr_start_time)
+                    logger.dump_tabular(with_prefix=False)
+                    if self.plot:
+                        if initialize_plot == 0:
+                            plotter.init_worker(sess)
+                            plotter.init_plot(self.env, self.policy)
+                            initialize_plot = 1
+                        plotter.update_plot(self.policy, self.max_path_length)
+                        if self.pause_for_plot:
+                            input("Plotting evaluation run: Press Enter to "
+                                  "continue...")
+        except KeyboardInterrupt:
```

Comment:
Is there a reason the KeyboardInterrupt code was added here?

Reply:
When we interrupt the parent process, cleanup does not happen properly for threads. So I handled the interrupt and made sure that shutdown_worker is called and cleanup happens properly.

Reply:
Okay, sounds good. Are there other interrupts which might stop the process?

Reply:
Apart from manually raising the interrupt, I didn't encounter any other possibility while executing both the algorithms.